File: stone_lib/src/main/java/com/github/klee0kai/stone/_hidden_/types/holders/MapItemHolder.java
package com.github.klee0kai.stone._hidden_.types.holders;

import com.github.klee0kai.stone._hidden_.types.ListUtils;
import com.github.klee0kai.stone._hidden_.types.SwitchCacheParam;
import com.github.klee0kai.stone.wrappers.Ref;

import java.lang.ref.Reference;
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Stone Private class
 */
public class MapItemHolder<Key, T> {

    private final StoneRefType defType;

    private StoneRefType curRefType;


    private final HashMap<Key, Object> refMap = new HashMap<>();
    private final AtomicInteger shedTaskCount = new AtomicInteger(0);


    public MapItemHolder(StoneRefType defType) {
        this.defType = defType;
        this.curRefType = defType;
    }

    public T get(Key key) {
        Object holder = refMap.get(key);
        if (holder == null) return null;
        switch (curRefType) {
            case StrongObject:
                return (T) holder;
            case WeakObject:
            case SoftObject:
                Reference<T> ref = ((Reference<T>) holder);
                return ref != null ? ref.get() : null;
        }
        return null;
    }

    public List<T> getList(Key key) {
        Object holder = refMap.get(key);
        if (holder == null) return null;
        switch (curRefType) {
            case ListObject:
                return (List<T>) holder;
            case ListWeakObject:
            case ListSoftObject:
                return ListUtils.format((List<Reference<T>>) holder, Reference::get);
        }
        return null;
    }

    public void set(Key key, Ref<T> creator, boolean onlyIfNull) {
        Object refHolder = refMap.get(key);
        if (Objects.equals(curRefType, StoneRefType.StrongObject)) {
            if (onlyIfNull && refHolder != null) return;
            refMap.put(key, creator.get());
            return;
        }
        ListUtils.IFormat<T, Reference<T>> formatter = curRefType.formatter();
        if (formatter == null) return;
        if (!onlyIfNull) {
            //switch ref type case
            refMap.put(key, formatter.format(creator.get()));
            return;
        }

        Reference<T> ref = (Reference<T>) refHolder;
        if (ref == null || ref.get() == null) {
            refMap.put(key, formatter.format(creator.get()));
        }
    }

    public void setList(Key key, Ref<List<T>> creator, boolean onlyIfNull) {
        Object refHolder = refMap.get(key);
        if (Objects.equals(curRefType, StoneRefType.ListObject)) {
            if (!onlyIfNull || refHolder == null) {
                refMap.put(key, creator.get());
                return;
            }

            // init nulls if needed
            List<T> created = null;
            List<T> refList = (List<T>) refHolder;
            for (int i = 0; i < refList.size(); i++) {
                if (refList.get(i) == null) {
                    if (created == null) created = creator.get();
                    refList.set(i, created.get(i));
                }
            }
            return;
        }
        ListUtils.IFormat<T, Reference<T>> formatter = curRefType.formatter();
        if (formatter == null) return;
        if (!onlyIfNull || refHolder == null) {
            //switch ref type case
            refMap.put(key, ListUtils.format(creator.get(), formatter));
            return;
        }

        List<Reference<T>> refList = (List<Reference<T>>) refHolder;
        List<T> created = null;
        for (int i = 0; i < refList.size(); i++) {
            if (refList.get(i) == null || refList.get(i).get() == null) {
                if (created == null) created = creator.get();
                refList.set(i, formatter.format(created.get(i)));
            }
        }
    }


    public void setRefType(StoneRefType refType) {
        if (curRefType == refType) return;
        if (defType.isList()) {
            HashMap<Key, List<T>> listMap = new HashMap<>();
            for (Key key : refMap.keySet()) listMap.put(key, getList(key));
            curRefType = refType.forList();
            for (Key key : listMap.keySet()) setList(key, () -> listMap.get(key), false);
        } else {
            HashMap<Key, T> itemMap = new HashMap<>();
            for (Key key : refMap.keySet()) itemMap.put(key, get(key));
            curRefType = refType.forSingle();
            for (Key key : itemMap.keySet()) set(key, () -> itemMap.get(key), false);
        }
    }


    public void remove(Key key) {
        refMap.remove(key);
    }


    public void reset() {
        curRefType = defType;
        refMap.clear();
    }


    public void clearNulls() {
        Set<Key> keys = new HashSet<>(refMap.keySet());
        if (!curRefType.isList()) {
            for (Key key : keys) {
                if (get(key) == null) {
                    refMap.remove(key);
                }
            }
        } else {
            for (Key key : keys) {
                List<T> list = getList(key);
                if (!ListUtils.contains(list, (i, it) -> it != null)) {
                    refMap.remove(key);
                }
            }
        }

    }


    public void switchCache(SwitchCacheParam args) {
        switch (args.cache) {
            case Default:
                setRefType(defType);
                return;
            case Reset:
                reset();
                return;
            case Weak:
                setRefType(StoneRefType.WeakObject);
                break;
            case Soft:
                setRefType(StoneRefType.SoftObject);
                break;
            case Strong:
                setRefType(StoneRefType.StrongObject);
                break;
        }

        if (args.time > 0) {
            shedTaskCount.incrementAndGet();
            args.scheduler.schedule(new ScheduleTask(args.time) {
                @Override
                public void run() {
                    if (shedTaskCount.decrementAndGet() <= 0) setRefType(defType);
                }
            });
        }
    }

}

File: stone_multiplatform/src/commonMain/kotlin/com/github/klee0kai/stone/__hidden__/types/holders/MapItemHolder.kt
package com.github.klee0kai.stone.__hidden__.types.holders

import com.github.klee0kai.stone.__hidden__.StoneScope
import com.github.klee0kai.stone.__hidden__.SwitchCacheParam
import com.github.klee0kai.stone.annotations.component.SwitchCache
import com.github.klee0kai.stone.weakref.Ref
import kotlinx.atomicfu.atomic
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Stone Private class
 */
@Suppress("UNCHECKED_CAST")
class MapItemHolder<Key, T>(
    private val defType: StoneRefType
) {

    private var curRefType: StoneRefType = defType


    private val refMap: HashMap<Key?, Any?> = HashMap()
    private val shedTaskCount = atomic(0)


    fun get(
        key: Key?,
    ): T? {
        val holder = refMap[key] ?: return null
        return when (curRefType) {
            StoneRefType.StrongObject -> holder as T?
            StoneRefType.WeakObject, StoneRefType.SoftObject -> (holder as Ref<T?>?)?.get()
            else -> null
        }
    }

    fun getList(
        key: Key?,
    ): List<T>? {
        val holder = refMap[key] ?: return null
        return when (curRefType) {
            StoneRefType.ListObject -> holder as MutableList<T>?
            StoneRefType.ListWeakObject, StoneRefType.ListSoftObject -> {
                (holder as MutableList<Ref<T?>?>?)?.mapNotNull { it?.get() }
            }

            else -> null
        }
    }

    fun getListNullable(
        key: Key?,
    ): List<T?>? {
        val holder = refMap[key] ?: return null
        return when (curRefType) {
            StoneRefType.ListObject -> holder as MutableList<T>?
            StoneRefType.ListWeakObject, StoneRefType.ListSoftObject -> {
                (holder as MutableList<Ref<T?>?>?)?.map { it?.get() }
            }

            else -> null
        }
    }

    fun set(
        key: Key?,
        onlyIfNull: Boolean,
        creator: Ref<T?>,
    ) {
        val refHolder = refMap[key]
        if (curRefType == StoneRefType.StrongObject) {
            if (onlyIfNull && refHolder != null) return
            refMap[key] = creator.get()
            return
        }
        val formatter: (T?) -> Ref<T?> = curRefType.formatter<T>() ?: return
        if (!onlyIfNull) {
            //switch ref type case
            refMap[key] = formatter(creator.get())
            return
        }

        val ref: Ref<T?>? = refHolder as Ref<T?>?
        if (ref == null || ref.get() == null) {
            refMap[key] = formatter(creator.get())
        }
    }

    fun setList(
        key: Key?,
        onlyIfNull: Boolean,
        creator: Ref<List<T?>?>,
    ) {
        val refHolder = refMap[key]
        if (curRefType == StoneRefType.ListObject) {
            if (!onlyIfNull || refHolder == null) {
                refMap[key] = creator.get()?.toMutableList()
                return
            }

            // init nulls if needed
            var created: List<T?>? = null
            val refList = refHolder as MutableList<T?>
            for (i in refList.indices) {
                if (refList[i] == null) {
                    if (created == null) created = creator.get()
                    refList[i] = created?.get(i)
                }
            }
            return
        }

        val formatter = curRefType.formatter<T>() ?: return
        val refList = refHolder as? MutableList<Ref<T?>?>
        if (!onlyIfNull || refList == null) {
            //switch ref type case
            refMap[key] = creator.get()?.map { formatter(it) }?.toMutableList()
            return
        }

        var created: List<T?>? = null
        for (i in refList.indices) {
            if (refList[i] == null || refList[i]?.get() == null) {
                if (created == null) created = creator.get()
                refList[i] = formatter(created?.get(i))
            }
        }
    }


    fun setRefType(refType: StoneRefType) {
        if (curRefType == refType) return
        if (defType.isList) {
            val listMap: HashMap<Key?, List<T?>?> = HashMap()
            for (key in refMap.keys) listMap[key] = getListNullable(key)
            curRefType = refType.forList()
            for (key in listMap.keys) setList(key, onlyIfNull = false) { listMap[key] }
        } else {
            val itemMap: HashMap<Key?, T?> = HashMap()
            for (key in refMap.keys) itemMap[key] = get(key)
            curRefType = refType.forSingle()
            for (key in itemMap.keys) set(key, onlyIfNull = false) { itemMap[key] }
        }
    }


    fun remove(key: Key?) {
        refMap.remove(key)
    }


    fun reset() {
        curRefType = defType
        refMap.clear()
    }


    fun clearNulls() {
        val keys: Set<Key?> = HashSet(refMap.keys)
        if (!curRefType.isList) {
            for (key in keys) {
                if (get(key) == null) {
                    refMap.remove(key)
                }
            }
        } else {
            for (key in keys) {
                val list = getListNullable(key)
                if (list?.none { it != null } == true) {
                    refMap.remove(key)
                }
            }
        }
    }


    fun switchCache(args: SwitchCacheParam) {
        when (args.cache) {
            SwitchCache.CacheType.Default -> {
                setRefType(defType)
                return
            }

            SwitchCache.CacheType.Reset -> {
                reset()
                return
            }

            SwitchCache.CacheType.Weak -> setRefType(StoneRefType.WeakObject)
            SwitchCache.CacheType.Soft -> setRefType(StoneRefType.SoftObject)
            SwitchCache.CacheType.Strong -> setRefType(StoneRefType.StrongObject)
        }

        if (args.time > 0) {
            shedTaskCount.incrementAndGet()
            StoneScope.stoneCoroutineScope.launch {
                delay(args.time)
                if (shedTaskCount.decrementAndGet() <= 0) {
                    setRefType(defType)
                }
            }
        }
    }
}
